src/xdp-utils.c

```diff
@@ -2163,16 +2364,99 @@ xdp_app_info_ensure_pidns (XdpAppInfo  *app_info,
     {
       int code = g_io_error_from_errno (-r);
       g_set_error (error, G_IO_ERROR, code,
-                   "Could not query /proc/%u/ns/pid: %s",
-                   (guint) pid, g_strerror (-r));
+                   "Could not query pidfd for pidns: %s",
+                   g_strerror (-r));
       return FALSE;
     }
 
-  app_info->u.flatpak.pidns_id = ns;
+  app_info->pidns_id = ns;
 
   return TRUE;
 }
 
+static gboolean
+xdp_app_info_ensure_pidns_host (XdpAppInfo  *app_info,
+                                DIR         *proc,
+                                GError     **error)
+{
+  ino_t ns;
+  int r;
+
+  if (app_info->u.host.pidfd < 0)
+    return FALSE;
+
+  r = lookup_ns_from_pid_fd (app_info->u.host.pidfd, &ns);
+  if (r < 0)
+    {
+      int code = g_io_error_from_errno (-r);
+      g_set_error (error, G_IO_ERROR, code,
+                   "Could not query pidfd for pidns: %s",
+                   g_strerror (-r));
+      return FALSE;
+    }
+
+  app_info->pidns_id = ns;
+  return TRUE;
+}
+
+static gboolean
+xdp_app_info_ensure_pidns_containers1 (XdpAppInfo  *app_info,
+                                       DIR         *proc,
+                                       GError     **error)
+{
+  ino_t ns;
+  int r;
+
+  if (xdp_app_info_is_flatpak (app_info))
+    {
+      /* Containers1 is supposed to be generic but currently flatpak still
+       * sets up the xdg-dbus-proxy which the pidfd is pointing at. When dbus
```

Comment:
> the container framework will need to tell us, in the metadata, that we can rely on that as an invariant

... which is another thing for which we will need somewhere where we can standardize the metadata.

Alternatively, x-d-p could pessimistically assume that the pid and pidfd might always refer to a proxy rather than the real sandboxed peer - which would decouple the pidfd stuff from the use of Containers1, making this PR a lot smaller and more manageable. I think that would be a win.

Reply:
In parallel, if you want to get uses of the pidfd in, you could teach x-d-p to assume that the pidfd is valid to use for HOST processes, but not for CONTAINERS1 or FLATPAK - that would let you implement the pidfd stuff in a separate PR that is not tightly coupled to this one.

I'm honestly not sure which way SNAP should go, you'd have to talk to a Snap developer. I don't think they use xdg-dbus-proxy *at the moment* (because they rely on Ubuntu's patched kernel being able to mediate D-Bus with AppArmor), but I also don't think we can assume as a load-bearing security mechanism that they *will never* use a filtering D-Bus proxy, unless someone has asked them and got a positive answer.